warehouse/schema/schema_v2.go

```diff
@@ -0,0 +1,217 @@
+	consolidatedSchema = enhanceSchemaWithIDResolution(consolidatedSchema, sh.isIDResolutionEnabled(), sh.warehouse.Type)
+
+	return consolidatedSchema, nil
+}
+
+func (sh *schemaV2) isIDResolutionEnabled() bool {
+	return sh.enableIDResolution && slices.Contains(whutils.IdentityEnabledWarehouses, sh.warehouse.Type)
+}
+
+func (sh *schemaV2) UpdateLocalSchemaWithWarehouse(ctx context.Context) error {
+	// no-op
+	return nil
+}
+
+func (sh *schemaV2) TableSchemaDiff(ctx context.Context, tableName string, tableSchema model.TableSchema) (whutils.TableSchemaDiff, error) {
+	schema, err := sh.getSchema(ctx)
+	if err != nil {
+		return whutils.TableSchemaDiff{}, fmt.Errorf("getting schema: %w", err)
+	}
+	return tableSchemaDiff(tableName, schema, tableSchema), nil
+}
+
+func (sh *schemaV2) FetchSchemaFromWarehouse(ctx context.Context, _ fetchSchemaRepo) error {
+	_, err := sh.fetchSchemaFromWarehouse(ctx)
+	return err
+}
+
+func (sh *schemaV2) fetchSchemaFromWarehouse(ctx context.Context) (model.Schema, error) {
+	warehouseSchema, err := sh.fetchSchemaRepo.FetchSchema(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("fetching schema: %w", err)
+	}
+	removeDeprecatedColumns(warehouseSchema, sh.warehouse, sh.log)
+	return warehouseSchema, sh.saveSchema(ctx, warehouseSchema)
+}
+
+func (sh *schemaV2) saveSchema(ctx context.Context, newSchema model.Schema) error {
+	expiresAt := sh.now().Add(sh.ttlInMinutes)
+	_, err := sh.schemaRepo.Insert(ctx, &model.WHSchema{
+		SourceID:        sh.warehouse.Source.ID,
+		Namespace:       sh.warehouse.Namespace,
+		DestinationID:   sh.warehouse.Destination.ID,
+		DestinationType: sh.warehouse.Type,
+		Schema:          newSchema,
+		ExpiresAt:       expiresAt,
+	})
+	if err != nil {
+		return fmt.Errorf("inserting schema: %w", err)
+	}
+	sh.cachedSchemaMu.Lock()
+	sh.cachedSchema = newSchema
+	sh.cachedSchemaMu.Unlock()
+	sh.cacheExpiry = expiresAt
+	return nil
+}
+
+func (sh *schemaV2) getSchema(ctx context.Context) (model.Schema, error) {
+	sh.cachedSchemaMu.RLock()
+	defer sh.cachedSchemaMu.RUnlock()
+	if sh.cachedSchema != nil && sh.cacheExpiry.After(sh.now()) {
+		return sh.cachedSchema, nil
+	}
+	whSchema, err := sh.schemaRepo.GetForNamespace(
+		ctx,
+		sh.warehouse.Source.ID,
+		sh.warehouse.Destination.ID,
+		sh.warehouse.Namespace,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting schema for namespace: %w", err)
+	}
+	if whSchema.Schema == nil {
+		sh.cachedSchema = model.Schema{}
+		sh.cacheExpiry = sh.now().Add(sh.ttlInMinutes)
+		return sh.cachedSchema, nil
+	}
+	if whSchema.ExpiresAt.Before(sh.now()) {
+		sh.log.Infon("Schema expired", obskit.DestinationID(sh.warehouse.Destination.ID), obskit.Namespace(sh.warehouse.Namespace), logger.NewTimeField("expiresAt", whSchema.ExpiresAt))
+		sh.cachedSchemaMu.RUnlock() // Release the lock since "fetchSchemaFromWarehouse" internally acquires a lock
+		schema, err := sh.fetchSchemaFromWarehouse(ctx)
```

Comment:
For a brief moment, there won't be any lock on the mutex. considering it is a read lock, Even if another reader locked it again, we would release it using `RUnlock`,  so writes in fetch will not be locked as there are no readers. But there is a chance there might be another reader acquiring read lock and there will be two `fetchSchemaFromWarehouse` calls?. Let me know if my understanding is right? is this even possible like two readers attempting to getSchema() concurrently?

Reply:
Good catch. Yes it is possible for two readers to call getSchema() concurrently. I've updated the code to apply the mutex selectively as needed.

Reply:
Offline discussion: we discussed the case of multiple readers acquiring the read locks and attempting to fetch schema from warehouse. we might have to acquire write lock early once the cache is missed or even at the start of the `getSchema` directly

Reply:
- Currently there is no scenario where the schemaV2 will be used concurrently. So it is theoretically fine to have multiple concurrent `fetchSchemaFromWarehouse` calls. The only way to prevent this would be to inline the function in `getSchema`